packages/editor/src/commands/ModifyPropertyCommand.ts

```diff
@@ -2,113 +2,126 @@ import Command, { CommandParams } from './Command'
-        const oldValue = res[propertyName]
+        for (const propertyName in params.properties) {
+          const { result, finalProp } = this.getNestedObject(comp, propertyName)
+          const oldValue = result[finalProp]
+          oldProps[propertyName] = oldValue && oldValue.clone ? oldValue.clone() : oldValue
+        }
 
-        objectOldProperties[propertyName] = oldValue && oldValue.clone ? oldValue.clone() : oldValue
+        this.oldProperties.push(oldProps)
       }
-
-      this.oldProperties.push(objectOldProperties)
     }
   }
 
   execute() {
-    this.updateProperties(this.affectedObjects, this.newProperties)
+    this.updateProperties(this.affectedObjects, this.properties, this.component)
+  }
+
+  shouldUpdate(newCommand: ModifyPropertyCommand): boolean {
+    return (
+      this.component === newCommand.component &&
+      arrayShallowEqual(Object.keys(this.properties), Object.keys(newCommand.properties)) &&
+      arrayShallowEqual(this.affectedObjects, newCommand.affectedObjects)
+    )
+  }
+
+  update(command: ModifyPropertyCommand) {
+    this.properties = command.properties
+    this.updateProperties(this.affectedObjects, command.properties, this.component)
   }
 
   undo() {
+    if (!this.oldProperties) return
+
     for (let i = 0; i < this.oldProperties.length; i++) {
-      this.updateProperties(this.affectedObjects, this.oldProperties[i])
+      this.updateProperties(this.affectedObjects, this.oldProperties[i], this.component)
     }
   }
 
   toString() {
     return `SetPropertiesMultipleCommand id: ${this.id} objects: ${serializeObject3DArray(
       this.affectedObjects
-    )} properties: ${serializeProperties(this.newProperties)}`
-  }
-
-  emitAfterExecuteEvent() {
-    if (this.shouldEmitEvent) {
-      CommandManager.instance.emitEvent(EditorEvents.OBJECTS_CHANGED, this.affectedObjects)
-    }
+    )} properties: ${serializeProperties(this.properties)}`
   }
 
-  updateProperties(objects?: any[], properties?: PropertyType): void {
-    for (const propertyName in properties) {
-      if (!Object.prototype.hasOwnProperty.call(properties, propertyName)) continue
-
-      const finalProp = propertyName.split('.').pop()
-
-      for (let i = 0; i < objects.length; i++) {
-        const object = objects[i]
-
-        const value = properties[propertyName]
-
-        const res = this.getNestedObject(object, propertyName)
-
-        if (value && value.copy) {
-          res[finalProp].copy(value)
-        } else {
-          res[finalProp] = value
-        }
-
-        if (object.onChange) {
-          object.onChange(propertyName)
+  updateProperties(nodes: EntityTreeNode[], properties: PropertyType, component: ComponentConstructor<any, any>): void {
+    for (let i = 0; i < nodes.length; i++) {
+      const entity = nodes[i].entity
+      if (component) {
+        const comp = getComponent(entity, component)
+        if (comp) {
+          for (const propertyName in properties) {
```

Comment:
avoid using the `for in` loop it is the slowest of them all. you can run a loop with Object.keys(properties)